main.ml

```diff
@@ -589,559 +597,577 @@ let process_smapiv2_requests ~volume_script_dir =
+      let vdi = Storage_interface.Vdi.string_of vdi_info.Storage_interface.vdi in
+      return_volume_rpc (fun () -> Volume_client.snapshot volume_rpc dbg sr vdi)
+      >>>= fun response ->
+      let now = Xapi_stdext_date.Date.(to_string (of_float (Unix.gettimeofday ()))) in
+      set ~dbg ~sr ~vdi:response.Xapi_storage.Control.key ~key:_snapshot_time_key ~value:now >>>= fun () ->
+      set ~dbg ~sr ~vdi:response.Xapi_storage.Control.key ~key:_is_a_snapshot_key ~value:(string_of_bool true) >>>= fun () ->
+      set ~dbg ~sr ~vdi:response.Xapi_storage.Control.key ~key:_snapshot_of_key ~value:vdi >>>= fun () ->
+      let response = { (vdi_of_volume response) with snapshot_time = now; is_a_snapshot = true; snapshot_of = (Storage_interface.Vdi.of_string vdi) } in
+      Deferred.Result.return response
+    end |> wrap
+  in
+  S.VDI.snapshot vdi_snapshot_impl;
+
+  let vdi_clone_impl dbg sr vdi_info =
+    begin
+      Attached_SRs.find sr >>>= fun sr ->
+      clone ~dbg ~sr ~vdi:(Storage_interface.Vdi.string_of vdi_info.Storage_interface.vdi)
+      >>>= fun response ->
+      Deferred.Result.return (vdi_of_volume response)
+    end |> wrap
+  in
+  S.VDI.clone vdi_clone_impl;
+
+  let vdi_set_name_label_impl dbg sr vdi' new_name_label =
+    begin
+      let vdi = Storage_interface.Vdi.string_of vdi' in
+      Attached_SRs.find sr >>>= fun sr ->
+      return_volume_rpc (fun () -> Volume_client.set_name volume_rpc dbg sr vdi new_name_label)
+    end |> wrap
+  in
+  S.VDI.set_name_label vdi_set_name_label_impl;
+
+  let vdi_set_name_description_impl dbg sr vdi' new_name_description =
+    begin
+      let vdi = Storage_interface.Vdi.string_of vdi' in
+      Attached_SRs.find sr >>>= fun sr ->
+      return_volume_rpc (fun () -> Volume_client.set_description volume_rpc dbg sr vdi new_name_description)
+    end |> wrap
+  in
+  S.VDI.set_name_description vdi_set_name_description_impl;
+
+  let vdi_resize_impl dbg sr vdi' new_size =
+    begin
+      let vdi = Storage_interface.Vdi.string_of vdi' in
+      Attached_SRs.find sr >>>= fun sr ->
+      return_volume_rpc (fun () -> Volume_client.resize volume_rpc dbg sr vdi new_size) >>>= fun () ->
+      (* Now call Volume.stat to discover the size *)
+      stat ~dbg ~sr ~vdi
+      >>>= fun response ->
+      Deferred.Result.return response.Xapi_storage.Control.virtual_size
+    end |> wrap
+  in
+  S.VDI.resize vdi_resize_impl;
+
+  let vdi_stat_impl dbg sr vdi' =
+    begin
+      let vdi = Storage_interface.Vdi.string_of vdi' in
+      Attached_SRs.find sr >>>= fun sr ->
+      stat ~dbg ~sr ~vdi >>>= fun response ->
+      Deferred.Result.return (vdi_of_volume response)
+    end |> wrap
+  in
+  S.VDI.stat vdi_stat_impl;
+
+  let vdi_introduce_impl dbg sr uuid sm_config location =
+    begin
+      Attached_SRs.find sr >>>= fun sr ->
+      let vdi = location in
+      stat ~dbg ~sr ~vdi >>>= fun response ->
+      Deferred.Result.return (vdi_of_volume response)
```

Comment:
I know this was like this in the old API, but how does stat-ing a VDI we're just about to introduce help the SR know that it should insert that VDI into its local tracking metadata (e.g. a sqlite db for gfs2)

Reply:
It's more that this is what xapi does when a VDI is introduced to _it_ rather than introducing a VDI to the backend. That is, when you do the XenAPI call 'VDI.introduce' it calls SMAPIv2 'VDI.introduce' to make sure the backend _already_ knows about it.

Reply:
Ah, got it, so the VDI must already exist on the backend.